### ISSUE TITLE
[react-hooks/exhaustive-deps] false positive for hook dependencies check

### ISSUE THREAD

**Issue by zdila** [typescript-eslint/typescript-eslint#2476]
After upgrading @typescript-eslint/eslint-plugin and @typescript-eslint/parser from 3.10.1 to 4.0.1 I am getting false positive for hook dependencies check.

```
WARNING in components/DrawingMenu.tsx:80:7
react-hooks/exhaustive-deps: React Hook useCallback has a complex expression in the dependency array. Extract it to a separate variable so it can be statically checked.
    78 |     },
    79 |     [
  > 80 |       selection?.id,
       |       ^^^^^^^^^^^^^
    81 |       selection?.type,
    82 |       linePoints,
    83 |       onInfoPointAdd,
```
Please let me know it it is problem of `eslint-plugin-react-hooks` so I will report it there. Thanks.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `12.15.0` |

**Comment by bradzacher** [comment]
This is a bug in their lint rule. It does not support the new AST.
Looks like it's been fixed but not released.

https://github.com/facebook/react/commit/1396e4a8f5646f35929883cbb449d2c83e7cbc79#diff-26e3db67655052fb708395a89179543b
